integration-tests/src/tests/client/features/multinode_stateless_validators.rs

```diff
@@ -0,0 +1,674 @@
+    // 2. We transition to a different epoch then prev_epoch_id. (so we run until the epoch id changes)
+    test.run_for(Duration::seconds(20));
+
+    // Capture the id of the epoch we will check for the correct validator information in assert_validator_info.
+    let prev_epoch_id = test.data[0].client.client.chain.head().unwrap().epoch_id;
+    assert_ne!(prev_epoch_id, initial_epoch_id);
+
+    test.run_until(
+        |data| data[0].client.client.chain.head().unwrap().epoch_id != prev_epoch_id,
+        Duration::seconds(EPOCH_LENGTH as i64),
+    );
+
+    // Check that the balances are correct.
+    for i in NUM_VALIDATORS..NUM_ACCOUNTS {
+        let account = &accounts[i];
+        assert_eq!(
+            test.data.query_balance(account),
+            *balances.get(account).unwrap(),
+            "Account balance mismatch for account {}",
+            account
+        );
+    }
+
+    for idx in 0..NUM_VALIDATORS {
+        test.data[idx].state_sync_dumper.stop();
+    }
+
+    // Check the validator information for the epoch with the prev_epoch_id.
+    assert_validator_info(&test.data[0].client.client, prev_epoch_id, initial_epoch_id, &accounts);
+
+    // Give the test a chance to finish off remaining events in the event loop, which can
+    // be important for properly shutting down the nodes.
+    test.shutdown_and_drain_remaining_events(Duration::seconds(20));
+}
+
+/// Returns the CurrentEpochValidatorInfo for each validator account for the given epoch id.
+fn get_current_validators(
+    client: &Client,
+    epoch_id: EpochId,
+) -> HashMap<AccountId, CurrentEpochValidatorInfo> {
+    client
+        .epoch_manager
+        .get_validator_info(ValidatorInfoIdentifier::EpochId(epoch_id))
+        .unwrap()
+        .current_validators
+        .iter()
+        .map(|v| (v.account_id.clone(), v.clone()))
+        .collect()
+}
+
+/// Asserts the following:
+/// 1. Block and chunk producers produce block and chunks and also validate chunk witnesses.
+/// 2. Chunk validators only validate chunk witnesses.
+/// 3. Stake of both the block/chunk producers and chunk validators increase (due to rewards).
+/// TODO: Assert on the specific reward amount, currently it only checks that some amount is rewarded.
+fn assert_validator_info(
+    client: &Client,
+    epoch_id: EpochId,
+    initial_epoch_id: EpochId,
+    accounts: &Vec<AccountId>,
+) {
+    let validator_to_info = get_current_validators(client, epoch_id);
+    let initial_validator_to_info = get_current_validators(client, initial_epoch_id);
+
+    // Check that block/chunk producers generate blocks/chunks and also endorse chunk witnesses.
+    for idx in 0..NUM_BLOCK_AND_CHUNK_PRODUCERS {
+        let account = &accounts[idx];
+        let validator_info = validator_to_info.get(account).unwrap();
+        assert!(validator_info.num_produced_blocks > 0);
+        assert!(validator_info.num_expected_blocks >= validator_info.num_produced_blocks);
```

Comment:
super nit: when we have assertions like
```
assert!(a > b);
assert!(c >= a);
```
I find it more readable to write
```
assert!(a > b);
assert!(a <= c);
```

Reply:
I guess you meant `assert!(a < b);`?

Reply:
Hmm no? `a = validator_info.num_produced_blocks, b = 0, c = validator_info.num_expected_blocks`

Reply:
ok got it.